### PR TITLE
merge search_result to android_ir

### PR DESCRIPTION
**Pull request by Henry8192**
# Description
Log Viewer now supports reading from generic logs as well as android logs:
- Change `initializeStream()` in `FourByteClpIrStreamProtocolDecoder.js` so that instead of throwing an error when receiving non-android ir stream, it would log warning messages and proceed to process it with generic ir stream.
- Change `FourByteClpIrStreamReader.js` to adapt getting verbosityIdx from both log4j and attributeTable.

# Validation performed
- Open `android_ir.zst` and `generic-16MB.zst` in the browser using "Select Log File", both logs can be opened and shown correctly in the editor.

### CODE REVIEW

File: src/Viewer/components/MenuBar/MenuBar.js [hunk 1]
```diff
@@ -1,14 +1,31 @@
...
+    ChevronDoubleRight,
+    ChevronLeft,
+    ChevronRight,
+    CloudArrowDown, Download,
+    FileText,
+    Folder,
+    Gear,
+    Keyboard,
+    Moon,
+    Sun, XCircle
```
Comment (junhaoliao) [inline src/Viewer/components/MenuBar/MenuBar.js]: put Xcircle on a new line

File: src/Viewer/services/decoder/FourByteClpIrStreamProtocolDecoder.js [hunk 2]
```diff
@@ -84,11 +97,13 @@ class FourByteClpIrStreamProtocolDecoder {
             throw new Error(`Incompatible protocol version: ${version}`);
         }
         const androidVersion = metadata[PROTOCOL.METADATA.ANDROID_BUILD_VERSION_KEY];
-        if (null === androidVersion) {
-            throw new Error("This viewer is specilized for CLP Android IR streams.");
+        if (null !== androidVersion && undefined !== androidVersion) {
+            this.initializeAttributeTable(metadata);
+        }else  {
+            console.log("This viewer is specilized for CLP Android IR streams.");
```
Comment (junhaoliao) [inline src/Viewer/services/decoder/FourByteClpIrStreamProtocolDecoder.js]: I believe now we can remove this `else{}` block

File: src/Viewer/services/decoder/FourByteClpIrStreamReader.js [hunk 3]
```diff
@@ -148,23 +148,30 @@ class FourByteClpIrStreamReader {
```
Comment (junhaoliao) [inline src/Viewer/services/decoder/FourByteClpIrStreamReader.js]: remove extra line

File: src/Viewer/services/decoder/FourByteClpIrStreamReader.js [hunk 4]
```diff
@@ -148,23 +148,30 @@ class FourByteClpIrStreamReader {
...
+
 
         // Get the offset before we output anything to the buffer
         const beginOffset = outputResizableBuffer.getLength();
 
         this._tokenDecoder.decodeTimestamp(outputResizableBuffer, timestamp);
-        outputResizableBuffer.push(
-            FourByteClpIrStreamReader.textEncoder.encode(formattedAttribute)
-        );
+        if (Object.keys(attributeTable).length > 0)
```
Comment (junhaoliao) [inline src/Viewer/services/decoder/FourByteClpIrStreamReader.js]: this._streamProtocolDecoder.getNumAttributes() > 0

File: src/Viewer/services/decoder/FourByteClpIrStreamReader.js [hunk 5]
```diff
@@ -148,23 +148,30 @@ class FourByteClpIrStreamReader {
...
-        const pid = this._attrPool.get(attributeTable.pid).get_int_val();
-        const tid = this._attrPool.get(attributeTable.tid).get_int_val();
-        const tag = this._attrPool.get(attributeTable.tag).get_str_val();
-        const priority = this._attrPool.get(attributeTable.priority).get_int_val();
         let formattedAttribute = " ";
-        formattedAttribute += String(pid).padStart(5, " ") + " ";
-        formattedAttribute += String(tid).padStart(5, " ") + " ";
-        formattedAttribute += this.androidPriorityToChar(priority) + " ";
-        formattedAttribute += String(tag).padEnd(8, " ") + ": ";
+        if (Object.keys(attributeTable).length > 0)
```
Comment (junhaoliao) [inline src/Viewer/services/decoder/FourByteClpIrStreamReader.js]: this._streamProtocolDecoder.getNumAttributes() > 0

File: src/Viewer/services/decoder/FourByteClpIrStreamReader.js [hunk 6]
```diff
@@ -281,11 +288,17 @@ class FourByteClpIrStreamReader {
...
-        const verbosityIx =
+        const attributeTable = this._streamProtocolDecoder.getAttributeTable();
+        
+        let verbosityIx = null;
+        if (undefined !== attributeTable.priority) {
+            verbosityIx =
             this._attrPool
                 .get(this._streamProtocolDecoder.getAttributeTable().priority)
                 .get_int_val();
+        } else{
```
Comment (junhaoliao) [inline src/Viewer/services/decoder/FourByteClpIrStreamReader.js]: add space after `else`

File: src/Viewer/services/decoder/FourByteClpIrStreamReader.js [hunk 7]
```diff
@@ -281,11 +288,17 @@ class FourByteClpIrStreamReader {
...
         // Read the logtype and timestamp present in every message
         this._streamProtocolDecoder.readLogtype(this._dataInputStream, tag, this._logtype);
-        // const verbosityIx = this._getLog4jVerbosity();
-        const verbosityIx =
+        const attributeTable = this._streamProtocolDecoder.getAttributeTable();
+        
+        let verbosityIx = null;
+        if (undefined !== attributeTable.priority) {
+            verbosityIx =
             this._attrPool
```
Comment (junhaoliao) [inline src/Viewer/services/decoder/FourByteClpIrStreamReader.js]: Indent these
```
            this._attrPool
                .get(this._streamProtocolDecoder.getAttributeTable().priority)
                .get_int_val();
```

**Review by junhaoliao — CHANGES_REQUESTED** [review]
Please help addressing the posted comments. Thanks

File: src/Viewer/components/MenuBar/DownloadHelper.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    this.getBlob = function () {
+        return blob;
+    };
+};
+
+const downloadCompressedFile = () => {
+    const link = document.createElement("a");
+
+    // this opens the link in a new tab,
+    //  which avoids interruption of uncompressed logs download
```
Comment (LinZhihao-723) [inline src/Viewer/components/MenuBar/DownloadHelper.js]: Do not write inline comments like this. The code should be self-explainable

File: src/Viewer/components/MenuBar/DownloadHelper.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    link.target = "_blank";
+
+    const urlParams = new URLSearchParams(window.location.search);
+    const filePath = urlParams.get("filePath");
+    try {
+        // this URL constructor should only succeed if "filePath"
+        //  has a "protocol" scheme like "http:"
```
Comment (LinZhihao-723) [inline src/Viewer/components/MenuBar/DownloadHelper.js]: Same above

File: src/Viewer/components/MenuBar/DownloadHelper.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        new URL(filePath);
+        link.href = filePath;
+    } catch (e) {
+        console.log(`Unable to construct URL object from "${filePath}": ${e}. 
+        Assuming the file is from current origin. `);
```
Comment (LinZhihao-723) [inline src/Viewer/components/MenuBar/DownloadHelper.js]: This is not the proper multi-line style.
```suggestion
        console.log(
            `Unable to construct URL object from "${filePath}": ${e}.`
        );
```
Also the last part is not necessary

File: src/Viewer/components/MenuBar/MenuBar.js [hunk 11]
```diff
@@ -167,6 +188,111 @@ export function MenuBar ({
...
+    };
+
+    const hasFilePath = () => {
+        const urlParams = new URLSearchParams(window.location.search);
+        return urlParams.get("filePath");
+    };
+
+    window.onbeforeunload = function (e) {
+        changeStateCallback(STATE_CHANGE_TYPE.stopDownload, null);
+        if (downloadWorker.current !== null) {
```
Comment (LinZhihao-723) [inline src/Viewer/components/MenuBar/MenuBar.js]:
```suggestion
        if (null !== downloadWorker.current) {
```
We should use Yoda condition: https://en.wikipedia.org/wiki/Yoda_conditions

File: src/Viewer/services/WorkerPool.js [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        }
+        return null;
+    }
+
+    /**
+     * Frees a provided worker from the pool.
+     *
+     * @param {Worker} _worker
+     */
+    freeWorker (_worker) {
```
Comment (LinZhihao-723) [inline src/Viewer/services/WorkerPool.js]: Why using `_` prefix for pass-in arguments?

File: src/Viewer/services/WorkerPool.js [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+                break;
+            }
+        }
+    }
+
+    /**
+     * Process queue by assigning tasks to workers.
+     */
+    processQueue () {
+        if (this.taskQueue.length > 0) {
```
Comment (LinZhihao-723) [inline src/Viewer/services/WorkerPool.js]: Yoda condition

File: src/Viewer/services/decoder/FileManager.js [hunk 14]
```diff
@@ -247,6 +281,39 @@ class FileManager {
...
+     * @param {number} page Page to be decoded.
+     * @param {number} pageSize Size of the page to be decoded.
+     * @private
+     */
+    _decodePageWithWorker (page, pageSize) {
+        const numEventsAtLevel = this._logEventOffsetsFiltered.length;
+
+        // Calculate where to start decoding from and how many events to decode
+        // On final page, the numberOfEvents is likely less than pageSize
+        const targetEvent = ((page-1) * pageSize);
```
Comment (LinZhihao-723) [inline src/Viewer/services/decoder/FileManager.js]:
```suggestion
        const targetEvent = ((page - 1) * pageSize);
```

File: src/Viewer/services/decoder/decodeWorker.js [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+            );
+        } catch (error) {
+            // Ignore EOF errors since we should still be able
+            // to print the decoded messages
+            if (error instanceof DataInputStreamEOFError) {
+                // TODO Give visual indication that the stream is truncated
+                console.error("Stream truncated.");
+            } else {
+                console.log("random error");
+                throw error;
```
Comment (LinZhihao-723) [inline src/Viewer/services/decoder/decodeWorker.js]: We should do some more meaningful logging here...

**Review by LinZhihao-723 — CHANGES_REQUESTED** [review]
Did a quick round over the coding style. There are many places that we should iterate and improve. Should be careful in terms of:
1. variable naming
2. comments and doc string
3. Yoda condition
I would highly recommend you to try `WebStorm` as the IDE. We have style configure file that could help you auto-format the js files.

File: src/Viewer/services/decoder/decodeWorker.js (hunk 15 above)
Comment (Henry8192) [reply src/Viewer/services/decoder/decodeWorker.js]: Should we add more cases for error catching? If so, what kind of error are we expecting?

**Pull request closed without merging**